README.md
# church1936
Lambda calculus implemented directly from Alonzo Church's 1936 paper “An Unsolvable Problem of Elementary Number Theory.” _American Journal of Mathematics_, vol. 58, no. 2, 1936, pp. 345–63. JSTOR, https://doi.org/10.2307/2371045.

[![Haskell CI](https://github.com/alfreb/church1936/actions/workflows/haskell.yml/badge.svg)](https://github.com/alfreb/church1936/actions/workflows/haskell.yml)

Build and run unit tests:
```
$ cabal test
```
Generate the documentation:
```
$ cabal haddock
```
Build and run the example program:
```
$ cabal run church1936
```

The example program prints the following (with some formatting added here for github readability): 

## A TL;DR of Church's paper

First the symbols of the lanugage of lambda calculus are introduced:\
`{}()𝝺,[]`\
\
Then the variables are introduced as a, b, c ... and so on. We're
told that this is an enumerably infinite set of symbols, but he
really only provides the first 3, so we have to improvise beyond z.
We did this by implementing Enum for variables as follows: \
`x, y, z, a₁, b₁, c₁, d₁, e₁, f₁, g₁`\
... and then \
`x₁, y₁, z₁, a₂, b₂, c₂, d₂, e₂, f₂, g₂` \
\
Any sequence of symbols defined above is a _formula_ in the paper
so technically `}(` and `{o₁[` are formulas, but only the
well-formed ones are interesting, which Church defines by induction
together with _free_ and _bound_ variables.  
\
"A variable *x* standing alone is a well-formed formula 
and the occurence of *x* in it is an occurence of *x* as a free
variable in it"\
\
Let's verify. x is a well formed formula: True
The `Variable` `x` is free in the `Formula` `Var (V "x")`: True
\
Further, if the formulas `F` and `X` are well-formed, `{F}(X)` is too.
This is one confirmed example: `{f}({f}(f))`\
\
Now for the lambda: if `M` is well formed and `x` is free in `M`, then
`x` is bound in `𝝺x[M]`. So 𝝺 has one job: to bind variables.
Here's one confirmed example of an `M` with `x` as free:\
`x(x(x(y)))`\
\
Attaching 𝝺x we get a well-formed formula with x now bound:\
`𝝺x . x(x(x(y)))`\
\
Church now defines the numerals as follows: 
```
1 → 𝝺ab . a(b)
2 → 𝝺ab . a(a(b))
3 → 𝝺ab . a(a(a(b)))
```
The successor function is not defined in this paper, but
was defined in Church 33 as follows:\
`𝝺cab . a(c(a,b))`\
\
We can apply the successor function to 2 like so:\
`{𝝺cab . a(c(a,b))}(𝝺ab . a(a(b)))`\
\
The full form of this expression is really this:\
`{𝝺c[𝝺a[𝝺b[{a}({{c}(a)}(b))]]]}(𝝺a[𝝺b[{a}({a}(b))]])`\
\
Thankfully, Church provided a nice set of abbreviation rules to
save us from the sea of brackets. You can turn them on or off.
\
Applying Operation II (reduction) repeatedly we eventually get
"normal form", where no further reduction is possible.
```
𝝺ab . a({𝝺ab . a(a(b))}(a,b))
𝝺ab . a({𝝺c . a(a(c))}(b))
𝝺ab . a(a(a(b)))
```
The successor of 2 is 3. Yay!\
\
Addition can now be implemented in terms of successor:\
`𝝺mn . n(𝝺cab . a(c(a,b)),m)`\
\
But for multiplication the plot thickens; Church messed up!
(see the documentation for details). Rosser helped fix this, as
noted in Kleene 35, with this nice baby:\
`𝝺abx . a(b(x))`\
\
And to top it off they point out that the numerals themselves
are exponentiation functions. 2(3) is 3^2 and 3(2) = 2^3.
Reversing the parameters we get it in the order we're used to:\
`𝝺ab . b(a)`\
\
Notice how this is exactly the same as the number 1, with the
variables in reversed order:\
`𝝺ab . a(b)`\
\
Let's end here, with proving that 2 + 2 = 4:\
`{𝝺pofx . p(f,o(f,x))}(𝝺ab . a(a(b)),𝝺ab . a(a(b)))`\
\
reduces to \
`𝝺fx . f(f(f(f(x))))`\
\
Pretty neat! 

## The point of the paper
Note that the objective of Church's paper was not to make a computer, it was to prove that Gödel's incompleteness theorem has implications for even simple statements in elementary number theory. More on that later!




CHANGELOG.md
# Revision history for church1936

## 0.1.0 -- 2023-02-23

* First working version. 


test/Spec.hs
--{-# OPTIONS_GHC -F -pgmF hspec-discover -optF --module-name=Spec #-}
{-# OPTIONS_GHC -F -pgmF hspec-discover #-} 


test/Church1936Spec.hs
module Church1936Spec (spec, testmain) where

import Test.Hspec
import Test.Hspec.QuickCheck
import Test.HUnit
import Test.QuickCheck

import Church1936 as C36

preamble = do
  putStrLn "Testing Church's Lambda calculus"


varExamples = [
  "a", "b", "c", "x", "y", "z",
  "a₁", "b₂", "c₉", "x₄₂","y₉₉", "z₇₈₉" ]

num0 = Lambda va (Lambda vb _b)
num4 = reduceFully $ Pair successor num3
snum1 = (Pair successor num1)
snum2 = (Pair successor snum1)

formula1 = (Lambda vx (Lambda vy (Pair _x _y)))

prop_numerals_valid (Positive n) = fromNumeral(toNumeral n) == n

-- Helper variables
va = var "a"
vb = var "b"
vc = var "c"
vd = var "d"
vf = var "f"
vi = var "i"
vk = var "k"
vm = var "m"
vn = var "n"
vx = var "x"
vy = var "y"
vz = var "z"


_a = toVar "a"
_b = toVar "b"
_c = toVar "c"
_n = toVar "n"
_m = toVar "m"
_x = toVar "x"
_y = toVar "y"
_z = toVar "z"


spec::Spec
spec = do
  describe "Variables" $ do
    describe "There are infinite variables, starting with 'a', 'b', 'c'" $ do
      it "Church doesnt't name the succesor of 'z' so we'll pick 'a₁'" $ do
        (succ $ var "z") `shouldBe` var "a₁"
      describe "The Enum class provides a total ordering for Variables:" $ do
        it "after a₁, b₁" $ do
          (succ $ succ $ var "z") `shouldBe` (var "b₁")
        it "after x₁, y₁" $ do
          (succ $ var "x₁") `shouldBe` var "y₁"
        it "after z₁, a₂" $ do
          (succ $ var "z₁") `shouldBe` var "a₂"
        it "after a₂, b₂" $ do
          (succ $ var "a₂") `shouldBe` var "b₂"
        it "after z₂, a₃" $ do
          (succ $ var "z₂") `shouldBe` var "a₃"
        it "z₁ < b₂" $ do
          (var "z₁") < (var "b₂") `shouldBe` True        
        it "z₁ > b₁" $ do
          (var "z₁") > (var "b₁") `shouldBe` True
          
        it "This also gives us a helpful list generator for variables" $ do
          [var "a"..] !! 26 `shouldBe` var "a₁"
          [var "a"..] !! (26 * 2) `shouldBe` var "a₂"
    
    it "Lowercase letters with and without subscript indices are variables" $ do
      (and $ map isVariableStr varExamples) `shouldBe` True
    
    it "Variables standing alone are free" $ do
      isFree vx (Var vx) `shouldBe` True

    it "x and y are free variables in the pair {x}(y)" $ do
      let f = Pair (toVar "x") (toVar "y") in
        isFree vx f
        && isFree vy f

    it "The free variables in 𝝺x . x(𝝺z . b(z)) are [b]" $ do 
      (freeVars $ Lambda vx (Pair _x (Lambda vz (Pair _b _z))))
        `shouldBe` [var "b"]
        
    it "x is bound in 𝝺x.x" $ do
      isFree vx (Lambda vx (Var vx)) `shouldBe` False
      isBound vx (Lambda vx (Var vx)) `shouldBe` True

    it "The bound vars in 𝝺ab . a(b) is [a,b]" $ do
      boundVars num1 `shouldBe` [var "a", var "b"]

    it "The bound vars in 𝝺ab . b is [b]" $ do
      boundVars (Lambda va (Lambda vb _b))
        `shouldBe` [var "b"]

    it "The bound vars in 𝝺ab . a is [a]" $ do
      boundVars (Lambda va (Lambda vb _a))
        `shouldBe` [var "a"] 

  describe "Formula abbreviation. " $ do
      it "We can generate abbreviated and non-abbreviated formulas" $ do
        abbreviate ON  formula1 `shouldBe` "𝝺xy . x(y)"
        abbreviate OFF formula1 `shouldBe` "𝝺x[𝝺y[{x}(y)]]"
        abbreviate OFF num1 `shouldBe` "𝝺a[𝝺b[{a}(b)]]" 
        abbreviate ON  num1 `shouldBe` "𝝺ab . a(b)"
        abbreviate OFF num2 `shouldBe` "𝝺a[𝝺b[{a}({a}(b))]]" 
        abbreviate ON  num2 `shouldBe` "𝝺ab . a(a(b))"  
        abbreviate OFF num3 `shouldBe` "𝝺a[𝝺b[{a}({a}({a}(b)))]]" 
        abbreviate ON  num3 `shouldBe` "𝝺ab . a(a(a(b)))"
        
  describe "Church numerals" $ do
    it "Are lambda's of two parameters, with nested pairs of variables" $ do
      Lambda va (Lambda vb (Pair _a _b)) `shouldBe` num1
      Lambda va (Lambda vb (Pair _a (Pair _a _b))) `shouldBe` num2
      Lambda va (Lambda vb (Pair _a (Pair _a (Pair _a _b)))) `shouldBe` num3

    it "The nesting is a right fold" $ do
      Lambda va (Lambda vb (foldr Pair _b [_a, _a, _a, _a, _a, _a, _a]))
        `shouldBe` toNumeral 7
      Lambda va (Lambda vb (foldl Pair _b [_a, _a, _a, _a, _a, _a, _a]))
        `shouldNotBe` toNumeral 7
    
    prop "Can be generated from numeric values and back to formulas"
      $ \(Positive n) -> fromNumeral(toNumeral n) `shouldBe` (n :: Int)
      
  describe "Successor" $ do
    it "Produces church numerals when redued" $ do
      (reduce $ reduce $ reduce snum1) `shouldBe` num2
      reduceFully snum2 `shouldBe` num3
      reduceFully (Pair successor num0) `shouldBe` num1

    it "Unreduced successor expressions are not equivalent to numerals" $ do
      Pair successor num0 `shouldNotBe` num1

  describe "Principal form" $ do
    it "Reductions often break equivalence due to variable renaming" $ do
      reduceFully (Pair (Pair add num1) num2) `shouldNotBe` num3
    it "Principal form harmonizes variable names to their \"natural order\"" $ do
      (principalForm $ reduceFully (Pair (Pair add num1) num2))
        `shouldBe` principalForm num3        
      (varsFromBinders $ principalForm (Pair (Pair add_church33 num2) num2))
        `shouldBe` [var "a" .. var "i"]
        
    it "Principal form of a pair of lambdas make bound variables unique" $ do
      (getVars $ principalForm $ Pair (Lambda vx _x) (Lambda vy _y))
        `shouldBe` [var "a", var "b"]
        
    it "Principal form of 𝝺xy . y(x) should be 𝝺ab . b(a)" $ do 
      (principalForm $ Lambda vx (Lambda vy (Pair _y _x)))
        `shouldBe` Lambda va (Lambda vb (Pair _b _a))
        
    it "Principal form of 𝝺x . x(𝝺z . b(z)) should be 𝝺a . a(𝝺c . b(c))" $ do
      (principalForm $ Lambda vx (Pair _x (Lambda vz (Pair _b _z))))
        `shouldBe` Lambda va (Pair _a (Lambda vc (Pair _b _c)))
        
    it "Principal form  of 𝝺ba . b(a) should be 𝝺ab . a(b)" $ do
      (principalForm (Lambda vb (Lambda va (Pair _b _a))))
        `shouldBe` (Lambda va (Lambda vb (Pair _a _b)))
       
    it "Principal form of 𝝺cab . a(c(a,b)) should be 𝝺abc . b(a(b,c))" $ do
      (principalForm successor) `shouldBe`
        Lambda va (Lambda vb (Lambda vc (Pair _b (Pair (Pair _a _b) _c))))
     
  describe "Principal Normal Form is examplified in Church 36 as follows" $ do
    it ("\"The formulas " ++ (show exp_rosser35)
        ++ " and 𝝺a . a(𝝺c . b(c)) are in principal normal form") $ do
      isPrincipalNormalForm exp_rosser35
        && isPrincipalNormalForm (Lambda va (Pair _a (Lambda vc (Pair _b _c))))
        `shouldBe` True
    it "and 𝝺ac . c(a), "$ do
      let f = Lambda va (Lambda vc (Pair _a _c))
        in
        isNormalForm f && not (isPrincipalNormalForm f) `shouldBe` True
    it "and 𝝺bc . c(b), " $ do
      let f = Lambda vb (Lambda vc (Pair _b _c))
        in isNormalForm f && not (isPrincipalNormalForm f) `shouldBe` True

    it ("and 𝝺a . a(𝝺a . b(a)) "
        ++ "are in normal form but not in principal normal form\"") $ do
      let f = Lambda va (Pair _a (Lambda va (Pair _b _a)))
        in isNormalForm f && not (isPrincipalNormalForm f) `shouldBe` True
  
  describe "Addition" $ do
      describe "Church defined addition in terms of successor and it works" $ do
        it "2 + 2 = 4" $ do
          (principalForm $ reduceFully (Pair (Pair add_church33 num2) num2))
            `shouldBe` principalForm num4
        it "4 + 4 = 8" $ do
          (principalForm $ reduceFully
            (Pair (Pair add_church33 $ toNumeral 4) (toNumeral 4)))
            `shouldBe` (principalForm $ toNumeral 8)
      it "17 + 23 = 40" $ do
        (principalForm $ reduceFully
         (Pair (Pair add_church33 $ toNumeral 17) (toNumeral 23)))
          `shouldBe` (principalForm $ toNumeral 40)

      describe "Rosser defined addition directly and that also works" $ do
        it "2 + 2 = 4" $ do
          (principalForm $ reduceFully (Pair (Pair add_rosser35 num2) num2))
            `shouldBe` principalForm num4
        it "4 + 4 = 8" $ do
          (principalForm $ reduceFully
            (Pair (Pair add_rosser35 $ toNumeral 4) (toNumeral 4)))
            `shouldBe` (principalForm $ toNumeral 8)
      it "17 + 23 = 40" $ do
        (principalForm $ reduceFully
         (Pair (Pair add_rosser35 $ toNumeral 17) (toNumeral 23)))
          `shouldBe` (principalForm $ toNumeral 40)

      it "The two implementations are not equivalent" $ do
        (principalForm $ reduceFully add_rosser35)
          `shouldNotBe` (principalForm $ reduceFully add_church33)
      

  describe "Multiplication" $ do
      describe "Church's definition relied on subtraction (Church 33)" $ do
        it "which relied on the iota operator, later found inconsistent" $ do
          False `shouldBe` False
      describe "Rosser's definition in Kleene 35 is direct and it works" $ do
        it "2 * 2 = 4" $ do
          (principalForm $ reduceFully (Pair (Pair mul_rosser35 num2) num2))
            `shouldBe` principalForm num4
        it "4 * 4 = 16" $ do
          (principalForm $ reduceFully
            (Pair (Pair mul_rosser35 $ toNumeral 4) (toNumeral 4)))
            `shouldBe` (principalForm $ toNumeral 16)
        it "17 * 23 = 391" $ do
          (principalForm $ reduceFully
           (Pair (Pair mul_rosser35 $ toNumeral 17) (toNumeral 23)))
            `shouldBe` (principalForm $ toNumeral 391)

      
  describe ("Exponentiation, not defined in Church 32-36, but in Kleene 35 as "
            ++ show exp_rosser35) $ do
    describe "Kleene 35 attributes this formulation to J.B. Rosser and it works" $ do
      it "2 ^ 2 = 4 with principal form applied frist" $ do
        (principalForm $ reduceFully $ principalForm (Pair (Pair exp_rosser35 num2) num2))
          `shouldBe` principalForm num4          
      it "2 ^ 2 = 4 without principal form applied first" $ do
        (principalForm $ reduceFully (Pair (Pair exp_rosser35 num2) num2))
          `shouldBe` principalForm num4
      it "4 ^ 4 = 256" $ do
        (fromNumeral $ reduceFully $ principalForm $
          app exp_rosser35 [toNumeral 4, toNumeral 4])
          `shouldBe` 256
          
      it "7 ^ 5 = 16807" $ do
        (fromNumeral $ reduceFully $ principalForm
          (Pair (Pair exp_rosser35 $ toNumeral 7) (toNumeral 5)))
          `shouldBe` 16807


testmain :: IO ()
testmain = hspec spec

tm = testmain
main = tm


test/HaskellNumeralsSpec.hs
module HaskellNumeralsSpec (spec, testmain) where

import Test.Hspec
import Test.Hspec.QuickCheck
import Test.HUnit
import Test.QuickCheck

import HaskellNumerals

preamble = do
  putStrLn "Testing Haskell Numerals"

-- =============================================================================
--                            Haskell Numerals
-- =============================================================================

haskellNumeral_8 = expH 2 3

rep1 f n = f n
rep2 f n = f(f n)
rep3 f n = f(f(f n))
rep5 f n = f(f(f(f(f n))))
rep7 f n = f(f(f(f(f(f(f n))))))

repz f x y = f(f x y)

nary_composes1 = rep1 repz 
nary_composes2 = repz . rep1

repz :: (t1 -> t2 -> t1) -> t1 -> t2 -> t2 -> t1


prop_compositionAffectsFirstOnly = ((-) . (^2) . (^2)) 2 2 == 14



n1 = numeralH 1
n2 = numeralH 2
n3 = numeralH 3
n4 = numeralH 4

-- Composing haskellNumerals gives us multiplication.
prop_composingHaskellNumeralsGivesMultiplication =
  (n3 . n2 . n3) (+1) 0 == 3 * 2 * 3 &&
  (n2 . n3 . n4) (+1) 0 == 3 * 2 * 4 &&
  (n4 . n4 . n4) (+1) 0 == 4 * 4 * 4


-- This is captured very nicely in the type signature of our haskellNumerals:
n3 :: (t -> t) -> t -> t


three = n3 succ 0
four  = n4 succ 0

twelve  = n4 (+ (n3 succ 0)) 0
twelve' = n3 (+ (n4 succ 0)) 0

perfect f v = f v

--
-- One numeral N applied to another, M becomes M
--
-- Because:
-- 
--(No, it's only when 1 is one of them)



-- Now
prop_haskellNumeralsAreExponentialFunctions =
  n2 n2 (+1) 0 == 2^2 && 
  n2 n3 (+1) 0 == 3^2 &&
  n2 n4 (+1) 0 == 4^2 &&
  n3 n2 (+1) 0 == 2^3 &&
  True

prop_haskellNumeralsFirstParamCanMultiply =
  n2 n2 (+2) 0 == (2^2) * 2 &&
  n2 n3 (+2) 0 == (3^2) * 2 &&
  n2 n3 (+3) 0 == (3^2) * 3 &&
  n2 n4 (+3) 0 == (4^2) * 3 &&
  True



spec::Spec
spec = do
  describe "Variables" $ do
    prop "Composing numerals is multiplication" $ do
      prop_composingHaskellNumeralsGivesMultiplication
        
    prop "Haskell numerals are exponential functions when applied to successor" $ do
      prop_haskellNumeralsAreExponentialFunctions

    prop "n-adder functions act as multipliers after exponentiation " $ do
      prop_haskellNumeralsFirstParamCanMultiply 

testmain :: IO ()
testmain = hspec spec

tm = testmain
main = tm


app/Main.hs
module Main where

import qualified Church1936 as C36

import Data.List

main :: IO ()
main = do
  putStrLn "\nA TL;DR of Church's paper.\n"
  putStrLn "First the symbols of the lanugage of lambda calculus are introduced:"
  putStrLn C36.symbols

  putStrLn "\nThen the variables are introduced as a, b, c ... and so on. We're"
  putStrLn "told that this is an enumerably infinite set of symbols, but he"
  putStrLn "really only provides the first 3, so we have to improvise beyond z."
  putStrLn "We did this by implementing Enum for variables as follows:"
  putStrLn $ (intercalate ", ") $ map show (take 10 [C36.var "x"..])
  putStrLn "... and then"
  putStrLn $ (intercalate ", ") $ map show (take 10 [C36.var "x₁"..])

  putStrLn "\nAny sequence of symbols defined above is a _formula_ in the paper"
  putStrLn "so technically `}(` and `{o₁[` are formulas, but only the"
  putStrLn "well-formed ones are interesting, which Church defines by induction"
  putStrLn "together with _free_ and _bound_ variables.\n"

  
  putStrLn "\"A variable *x* standing alone is a well-formed formula "
  putStrLn "and the occurence of *x* in it is an occurence of *x* as a free"
  putStrLn "variable in it\""

  putStr "\nLet's verify. x is a well formed formula: "
  putStrLn $ show $ C36.isWellFormed $ (C36.Var . C36.var) "x"
  putStr "The `Variable` `x` is free in the `Formula` `Var (V \"x\")`: "
  putStrLn $ show $ C36.isFree (C36.var "x") ((C36.Var . C36.var) "x")


  putStrLn "\nFurther, if the formulas F and X are well-formed, {F}(X) is too."

  let f = C36.toVar "f"
      x = C36.Pair f f
    in
    if C36.isWellFormed $ C36.Pair f x
    then
      putStrLn $ "This is one confirmed example: "
      ++ (C36.abbreviate C36.OFF $ C36.Pair f x)
    else error "Church would not have liked this implementation."

  putStrLn "\nNow for the lambda: if M is well formed and x is free in M, then"
  putStrLn "x is bound in 𝝺x[M]. So 𝝺 has one job: to bind variables."

  let x = C36.toVar "x"
      y = C36.toVar "y"
      m = C36.Pair x (C36.Pair x (C36.Pair x y))
    in
    if C36.isWellFormed m
       && C36.isFree (C36.var "x") m
    then
      putStrLn ("Here's one confirmed example of an M with x as free:\n"
                ++ show m)
    else
      error "Church would not approve of this"
  
  putStrLn "\nAttaching 𝝺x we get a well-formed formula with x now bound:"
  let x = C36.toVar "x"
      y = C36.toVar "y"
      m = C36.Pair x (C36.Pair x (C36.Pair x y))
      m'= C36.Lambda (C36.var "x") m
    in
      if C36.isWellFormed m'
         && C36.isBound (C36.var "x") m'
      then
        putStrLn $ (show m')
      else
        error "Failing this would be a disgrace to lambda calculus"

  
  putStrLn "\nChurch now defines the numerals as follows: "
  putStrLn ("1 → " ++ show C36.num1)
  putStrLn ("2 → " ++ show C36.num2)
  putStrLn ("3 → " ++ show C36.num3)
  
  putStrLn "\nThe successor function is not defined in this paper, but"
  putStrLn "was defined in Church 33 as follows:"
  putStrLn $ show C36.successor

  putStrLn "\nWe can apply the successor function to 2 like so:"
  putStrLn $ show $ C36.Pair C36.successor C36.num2

  putStrLn "\nThe full form of this expression is really this:"
  putStrLn $ C36.abbreviate C36.OFF (C36.Pair C36.successor C36.num2)

  putStrLn "\nThankfully, Church provided a nice set of abbreviation rules to"
  putStrLn "save us from the sea of brackets. You can turn them on or off."
  
  putStrLn "\nApplying Operation II (reduction) repeatedly we eventually get"
  putStrLn "\"normal form\", where no further reduction is possible."
  putStrLn $ show $ C36.reduceN 1 (C36.Pair C36.successor C36.num2) 
  putStrLn $ show $ C36.reduceN 2 (C36.Pair C36.successor C36.num2)
  putStrLn $ show $ C36.reduceN 3 (C36.Pair C36.successor C36.num2)
 
  putStrLn "\nThe successor of 2 is 3. Yay! "
  
  putStrLn "\nAddition can now be implemented in terms of successor:"
  putStrLn $ show $ C36.add_church33

  putStrLn "\nBut for multiplication the plot thickens; Church messed up!"
  putStrLn "(see the documentation for details). Rosser helped fix this, as"
  putStrLn "noted in Kleene 35, with this nice baby:"
  putStrLn $ show $ C36.mul_rosser35

  putStrLn "\nAnd to top it off they point out that the numerals themselves"
  putStrLn "are exponentiation functions. 2(3) is 3^2 and 3(2) = 2^3."
  putStrLn "Reversing the parameters we get it in the order we're used to:"
  putStrLn $ show $ C36.exp_rosser35
  
  putStrLn "\nNotice how this is exactly the same as the number 1, with the"
  putStrLn "variables in reversed order:"
  putStrLn $ show $ C36.num1


  putStrLn "\nLet's end here, with proving that 2 + 2 = 4:"
  putStrLn $ show $ C36.app C36.add [C36.num2, C36.num2]
  putStrLn "\nreduces to "
  putStrLn $ show $ C36.reduceFully $ C36.app C36.add [C36.num2, C36.num2]

  putStrLn "\nPretty neat! Check out the documentation and unit tests for more."
